airbyte-webapp/src/views/Connector/ServiceForm/useAuthentication.tsx

```diff
@@ -0,0 +1,210 @@
+   */
+  isHiddenAuthField: (fieldPath: string) => boolean;
+  /**
+   * A record of all formik errors in hidden authentication fields. The key will be the
+   * name of the field and the value an error code. If no error is present in a field
+   * it will be missing from this object.
+   */
+  hiddenAuthFieldErrors: Record<string, string>;
+  /**
+   * This will return true if the auth button should be visible and rendered in the place of
+   * the passed in field, and false otherwise.
+   */
+  shouldShowAuthButton: (fieldPath: string) => boolean;
+}
+
+export const useAuthentication = (): AuthenticationHook => {
+  const { values, getFieldMeta, submitCount } = useFormikContext<ServiceFormValues>();
+  const { selectedConnector } = useServiceForm();
+
+  const allowOAuthConnector = useFeature(FeatureItem.AllowOAuthConnector);
+
+  const advancedAuth = selectedConnector?.advancedAuth;
+  const legacyOauthSpec = selectedConnector?.authSpecification?.oauth2Specification;
+
+  const spec = selectedConnector?.connectionSpecification as JSONSchema7;
+
+  const isAuthButtonVisible = useMemo(() => {
+    const shouldShowAdvancedAuth =
+      advancedAuth && shouldShowButtonForAdvancedAuth(advancedAuth.predicateKey, advancedAuth.predicateValue, values);
+    const shouldShowLegacyAuth =
+      legacyOauthSpec && shouldShowButtonForLegacyAuth(spec, legacyOauthSpec.rootObject as Path, values);
+    return Boolean(allowOAuthConnector && (shouldShowAdvancedAuth || shouldShowLegacyAuth));
+  }, [values, advancedAuth, legacyOauthSpec, spec, allowOAuthConnector]);
+
+  // Fields that are filled by the OAuth flow and thus won't need to be shown in the UI if OAuth is available
+  const implicitAuthFieldPaths = useMemo(
+    () => [
+      // Fields from `advancedAuth` connectors
+      ...(advancedAuth
+        ? Object.values(serverProvidedOauthPaths(selectedConnector)).map((f) =>
+            makeConnectionConfigurationPath(f.path_in_connector_config)
+          )
+        : []),
+      // Fields from legacy `authSpecification` connectors
+      ...(legacyOauthSpec
+        ? [
+            ...convertAndPrefixPaths(legacyOauthSpec.oauthFlowInitParameters, legacyOauthSpec.rootObject as Path),
+            ...convertAndPrefixPaths(legacyOauthSpec.oauthFlowOutputParameters, legacyOauthSpec.rootObject as Path),
+          ]
+        : []),
+    ],
+    [advancedAuth, legacyOauthSpec, selectedConnector]
+  );
+
+  const isHiddenAuthField = useCallback(
+    (fieldPath: string) => {
+      // A field should be hidden due to OAuth if we have OAuth enabled and selected (in case it's inside a oneOf)
+      // and the field is part of the OAuth flow parameters.
+      return isAuthButtonVisible && implicitAuthFieldPaths.includes(fieldPath);
+    },
+    [implicitAuthFieldPaths, isAuthButtonVisible]
+  );
+
+  const hiddenAuthFieldErrors = useMemo(() => {
+    if (!isAuthButtonVisible) {
+      // We don't want to return the errors if the auth button isn't visible.
```

Comment:
I'm interested how we might end up with errors on these fields if the button isn't visible.  Is this because if a user is using a different auth type in a oneOf we still get back an error for the OAuth fields?  That seems odd.

Or if a user has an error with OAuth and then chooses another auth method the errors persist?  That also seems odd.

Reply:
In practise I don't think this happens at the moment, since we change our validation schema when the user actually changes the option of the `oneOf`, the "hidden" fields of the non-selected oneOf option won't be actually validated anymore and thus can't throw errors.

There is one case though why we need to filter out like that: the non OAuth option under the `oneOf` could have a field with the same name as one of the OAuth fields under the OAuth option, which would cause problems, since the actual field names, would not contain the index of the `oneOf`. E.g. if we imagine the following (pseudo) connector spec:

```
- credentials
  - oneOf
    - type: oauth
      properties:
      - token
      - client_id
    - type: token
      properties:
      - token
```

If the user would now select token in the oneOf and that `token` field is required there, but empty, we'd detect that as an OAuth error, because the OAuth field name `credentials.token` matches the same field name (since it doesn't have any notion of the selected oneOf in the field name). In theory that would still not cause a problem having this inside the `authErrors`, since the AuthButton itself (which currently is the only one using it) will never be rendered as long as the wrong option is selected, but I found it cleaner to not have this error in the `authErrors`, despite not being used.